README
# Computer-Security-Final course project

Name - Kartik Deshpande
Bmail ID - kdeshpa3

- Coded in python
- Code was compiled and tested on remote machines
- Used python's Crypto module to implement encryption and decryption
- Used python's socket server to implement concurrent server
-Execution 

./vf PORTNUM
./voter-cli IP-ADDR PORTNUM


db/history
114660000 	 Tue May  1 19:26:32 2018
113880000 	 Tue May  1 19:26:14 2018
112550000 	 Tue May  1 19:25:48 2018


test.py
import rsa
from base64 import b64encode, b64decode
import textwrap

msg1 = "Hello Tony, I am Jarvis!"
msg2 = "Hello Toni, I am Jarvis!"
keysize = 2048
(public, private) = rsa.newkeys(keysize)
encrypted = b64encode(rsa.encrypt(msg1, public))
decrypted = rsa.decrypt(b64decode(encrypted), private)
signature = b64encode(rsa.sign(msg1, private, "SHA-512"))
verify = rsa.verify(msg1, b64decode(signature), public)

print(private.exportKey('PEM'))
print(public.exportKey('PEM'))
print("Encrypted: " + encrypted)
print("Decrypted: '%s'" % decrypted)
print("Signature: " + signature)
print("Verify: %s" % verify)
rsa.verify(msg2, b64decode(signature), public)


print "#######################################"
name = "Alice"
Regno = "21313123"
#load public key
f = open("certs/voter_server_public",'r')                                                        
public_key = rsa.importKey(f.read())
signature = b64encode(rsa.sign(name, private, "SHA-512"))

full_encrypted_msg = ""

# add name 
full_encrypted_msg = b64encode(rsa.encrypt(name, public)) + b64encode(rsa.encrypt(Regno, public)) + signature

print full_encrypted_msg

print full_encrypted_msg[0:344]

print full_encrypted_msg[344:344+344]

print full_encrypted_msg[344+344:len(full_encrypted_msg)]


# decryption

decrypted = rsa.decrypt(b64decode(full_encrypted_msg[344:344+344]), private)
print decrypted


Socket_Sender.py
#!/usr/bin/python


import sys
import socket

MAX_REQUEST_SIZE = 10000

class Socket_Sender:

    # sending messages to server / clients
    def send_message(self, ip, port_num, message):
        try:
		message = message.replace("\r\n", '')
                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                sock.connect((ip, int(port_num)))
                sock.sendall(message)
                server_response = sock.recv(MAX_REQUEST_SIZE)
                sock.close()
                if server_response == "" :
                        # ERROR ! Empty response, probably the server does not want to send anything
                        return None
                else:
			server_response = server_response.replace("\r\n", '')
                        return server_response
        except Exception as e:
                print "ERROR ! Socket exception in send_message Socket Sender while sending message to " + str(ip) + " " + str(port_num)
		print e
		return None


voting_client.py
#!/usr/bin/python

import os
import time
import sys
import socket
import re

import rsa
from Socket_Sender import *
from Voting_Crypto import *
from Electorial_Voting import *

###########################
# stage 0 : Auth voting client with server
# stage 1 : Voting someone ( election ), ( check if voter has voted )
# stage 2 : Vote history
# stage 3 : Election result
# stage 4 : Quit 
###########################

class Client_Instance:
	sender = Socket_Sender()	# this will handle sending and receiving messages, socket
	crypto = Voting_Crypto()
	elec_vote = Electorial_Voting()

	voter_file = "votercli"
	public_cl = None
	private_cl = None
	public_vf = None
	private_vf = None
	response_success_status = "0"
	response_failure_status = "1"
	

	HOST = None
	PORT = None

	def __init__(self, host,port):
		self.HOST = host
		self.PORT = int(port)
		print "Connected to server " + str(host) + ":" + str(port)
		try:
			self.public_vf = self.crypto.load_public("voter_server_public")
			self.private_vf = self.crypto.load_private("voter_server_private")
		except Exception as e:
			print e
		self.elec_vote.initialize(self.voter_file)

	def auth_voting_client(self):

		stage = 0

		# console for user input
		while True:

        		name = raw_input("\nPlease enter your name\n")
        		regno = raw_input("\nPlease enter your registration number\n")

		        if name != "" and regno != "" :
				try:
					self.public_cl = self.crypto.load_public(name + "_" + regno + "_" + "public")
					self.private_cl = self.crypto.load_private(name + "_" + regno + "_" + "private")
					
					ciphertext = self.crypto.encrypt_message(self.public_vf,self.private_cl,name,regno,stage,"")

				except Exception as e:
					print "Invalid userID and registration number"
					#print "Private key of user " + name + " not found"
					#print e
					sys.exit(0)
		
				if ciphertext == False:
					print "Invalid userID and registration number"
					sys.exit(0)	
				else:
                			return_message = self.sender.send_message(self.HOST, self.PORT, ciphertext)
					if return_message == self.response_success_status :
						break
					else:
						print "Invalid userID and registration number"
						sys.exit(0)

		self.main_menu(name,regno)

	def main_menu(self,name,regno):
 		
		while True:
			print "\n"
			print "Welcome " + name	
			print "Main menu"
			print "Please enter a number (1-4)"
			print "1. Vote"
			print "2. My vote history"
			print "3. Election result"
			print "4. Quit"
			print "\n"
	
			userinput = raw_input()
			userinput = userinput.strip()

			try: 
				userinput = int(userinput)
			except:
				print "Invalid choice"

			if userinput != "" :
	
				try:
	                                userinput = int(userinput)
	                        except:
	                                print "Invalid choice"

				if userinput == 1:
					stage = 1
					extension = ""
					ciphertext = self.crypto.encrypt_message(self.public_vf,self.private_cl,name,regno,stage,extension)
				
					return_message = self.sender.send_message(self.HOST, self.PORT, ciphertext)
	                                if return_message == self.response_success_status:
						self.electorate(name,regno)
					else:
	                                        print "ERROR ! You have already voted"
	
				elif userinput == 2:
					stage = 2			
					extension = ""
					ciphertext = self.crypto.encrypt_message(self.public_vf,self.private_cl,name,regno,stage,extension)
				
	                                return_message = self.sender.send_message(self.HOST, self.PORT, ciphertext)
	                                if return_message == self.response_failure_status or return_message == "":
						print "No voter has voted yet."
					else:
						print return_message

				elif userinput == 3:
					stage = 3		
					extension = ""
	                                ciphertext = self.crypto.encrypt_message(self.public_vf,self.private_cl,name,regno,stage,extension)
					
					return_message = self.sender.send_message(self.HOST, self.PORT, ciphertext)
                                        if return_message == self.response_failure_status or return_message == "":
                                                print "Voting has not finished yet"
                                        else:
                                                print return_message

				elif userinput == 4:
                	                print "Good Bye"
                	                sys.exit(0)
				else:
					print "Invalid choice"
								
	def electorate(self,name,regno):
		while True :
			stage = 11
			
			print "Please enter a number ( 1 - " + str(len(self.elec_vote.election_database) + 2) + " ) "
			# get all election candidates to display
			for candidate in self.elec_vote.election_database:
				print str(self.elec_vote.election_database[candidate][0]) + ". " + candidate 
				last_index = self.elec_vote.election_database[candidate][0]
				back_id = last_index + 1
				quit_id = back_id + 1
			print str(back_id) + ". Back "
			print str(quit_id) + ". Quit" 
			print "\n"		

			userinput = raw_input()
		

			if userinput != "" :	
				try:
	                                userinput = int(userinput)
	                        except:
	                                print "Invalid choice"
	
				if userinput > quit_id or userinput < 1 :
					print "Invalid choice"
				elif userinput == back_id:
					break
				elif back_id == quit_id :
					print "Good Bye"
	                                sys.exit(0)
				else:
					stage = 11
					extension = userinput
					ciphertext = self.crypto.encrypt_message(self.public_vf,self.private_cl,name,regno,stage,extension)
	                                return_message = self.sender.send_message(self.HOST, self.PORT, ciphertext)
	                                if return_message == self.response_success_status :
                                                print "Vote Registration Success"
                                                break
					else:
	                                        print "Something went wrong, while registering your vote"
	
if len(sys.argv) != 3:
        print "Invalid Parameters: <VotingServer IP> <Port>"
        sys.exit(0)
else:
        HOST = sys.argv[1]
        PORT = sys.argv[2]

if __name__ == "__main__":
	client = Client_Instance(HOST,PORT)
	client.auth_voting_client()


Socket_Receive.py
#!/usr/bin/python

import os
import time
import sys
import SocketServer
import socket
from threading import Lock,Thread

import rsa
from Socket_Sender import *
from Electorial_Voting import *
from Voting_Crypto import *

MAX_REQUEST_SIZE = 10000

class Socket_Receive(SocketServer.ThreadingTCPServer):
        def __init__(self, server_address, RequestHandlerClass, FILENAME):
		SocketServer.ThreadingTCPServer.__init__(self,server_address,RequestHandlerClass)
		self.votingfile = FILENAME

###########################
# stage 0 : Auth voting client with server
# stage 1 : Voting someone ( election ), ( check if voter has voted )
#	    stage 11 : Vote a candidate ( A or B )
# stage 2 : Vote history
# stage 3 : Election result
# stage 4 : Quit
###########################

class Handle_Receive(SocketServer.BaseRequestHandler):
	
	elec_vote = Electorial_Voting()
	crypto = Voting_Crypto()

	voter_file = "votercli"
	private_vf = None
	public_vf = None


	# handle incoming socket requests from clients
    	def handle(self):
		if self.elec_vote.initialization_status == False:
			#self.initialize(self.server.votingfile)
			self.elec_vote.initialize(self.voter_file)

		# init private and public key
		#public_vf = self.crypto.load_public("voter_server_public")
              	private_vf = self.crypto.load_private("voter_server_private")

        	cipher = self.request.recv(MAX_REQUEST_SIZE)

		# TODO : Remove this part later	
		####################################
                obj = Voting_Crypto ()
                name = "Bob"
                reg_no = "113880000"
                public_vf = obj.load_public("voter_server_public")
                private_vf = obj.load_private("voter_server_private")
                private_cl = obj.load_private(name + "_" + reg_no + "_" + "private")
                cip =  obj.encrypt_message(public_vf, private_cl, name, reg_no,"0","")
		####################################

		# decrypt this message
		plaintext = self.crypto.decrypt_message(private_vf,cipher)

		# Print both encrypted and decrypted text
		print "Encrypted cipher text : "
		print cipher
		print "Decrypted plain text : "
		print plaintext

		status = plaintext[0]
		name = plaintext[1].strip()
		reg_no = plaintext[2].strip()
		stage = int(plaintext[3].strip())
		extension = plaintext[4].strip()

		respnse_success_status = "0"
		resonse_failure_status = "1"

		#print self.elec_vote.voter_database
		#print self.elec_vote.election_database

		if stage == 0:
			# auth user
			if self.elec_vote.validate_voter(name,reg_no) == False or status == False:
				# Not found in DB or signature verification failed
				self.request.send(resonse_failure_status)	
			else:
				self.request.send(respnse_success_status)	
		elif stage == 1 :
			# check if client has already voted
			if self.elec_vote.hasVoted(reg_no) == True:
				# he already has voted, can't change that
				self.request.send(resonse_failure_status)		
			else:
				self.request.send(respnse_success_status)	
		elif stage == 11 :
			# voting a particular candidate
			extension = int(extension)
			if self.elec_vote.electionVote(reg_no,extension) == False:
				self.request.send(resonse_failure_status)
			else:
				self.request.send(respnse_success_status)

			# check if all candidates have voted, if yes print the result
			if self.elec_vote.get_election_result() != "False":
				print self.elec_vote.get_election_result()
	
		elif stage == 2 :
			# voter history
			
			if self.elec_vote.get_voter_history(reg_no) != "False":
				response = self.elec_vote.get_voter_history(reg_no)
				response = self.format_string(response)
			else:
				response = resonse_failure_status
			self.request.send(response)

		elif stage == 3 :
			# election result

			if self.elec_vote.get_election_result() != "False":
                                response =  self.elec_vote.get_election_result()
				response = self.format_string(response)
			else:
				response = resonse_failure_status
			self.request.send(response)

		else:
			# Invalid 
			print "EXCEPTION ! This stage was not supposed to be considered " + stage


	def format_string(self,inmsg):
		return "######################################### \n" + inmsg + "#########################################"

	


generate-certs.py
import rsa
from base64 import b64encode, b64decode
import os
import shutil

KEYSIZE_DEFAULT = 2048

class GnerateCerts:

	keysize = KEYSIZE_DEFAULT
	file_votercli = None
	CERT_DIR = "certs/"

	def __init__(self, keysize_arg, filename_arg):
		self.keysize = keysize_arg
		self.file_votercli = filename_arg

	def write_tofile(self, private_key, public_key, name_of_file):
		
		try:
			# write private key
			filename = self.CERT_DIR +  name_of_file + "_private"
			priv_file_handle = open(filename, "w+")
			priv_file_handle.write(private_key.exportKey('PEM'))
			priv_file_handle.close()
		except IOError:
	                print "ERROR! The file " + filename + " was not found!"
			return			

		try:
			# write public key
			filename = self.CERT_DIR + name_of_file + "_public"
			pub_file_handle = open(filename, "w+")
	                pub_file_handle.write(public_key.exportKey('PEM'))
			pub_file_handle.close()
		except IOError:
                        print "ERROR! The file " + filename + " was not found!"
                        return


	def gnerate_cert(self):

		if os.path.isdir(self.CERT_DIR):
	
			# first empty all cert files in dir
			shutil.rmtree(self.CERT_DIR)

			# create the directory
                        os.makedirs(self.CERT_DIR)

		else:
			# create the directory
			os.makedirs(self.CERT_DIR)			

		# Generate Voter Server certs first
		(public, private) = rsa.newkeys(self.keysize)

		self.write_tofile(private,public,"voter_server")

		# generate public private keys for each voter
		if os.path.exists(self.file_votercli)  :
			file_handle = open(self.file_votercli, "r")
			for line in file_handle:
				line = line.rstrip()
				if line.strip() != "":
					a_line = line.split(' ')

					(public, private) = rsa.newkeys(self.keysize)
			                self.write_tofile(private,public,a_line[0] + "_" + a_line[1])
if __name__ == '__main__':
	cer = GnerateCerts(2048, "votercli")
	cer.gnerate_cert()


Electorial_Voting.py
#!/usr/bin/python

import os
import time
import sys
import SocketServer
import socket
from threading import Lock,Thread
import time

import rsa
from Socket_Sender import *



class Electorial_Voting:

        voter_file = "votercli"
        initialization_status = False
        voter_database = {}
        election_database = {}
        election_candidates = ['Tim' , 'Linda']
        CERT_DIR = "certs/"
	
	persistent_storage = "db/"
	result_file = "result"
	history_file = "history"
	write_to_storage = True

        def initialize(self,voterfile_arg):

                self.initialization_status = True

                self.voter_file = voterfile_arg

                # store all voter_info in local memory
                if os.path.exists(self.voter_file)  :
                        file_handle = open(self.voter_file, "r")
                        for line in file_handle:
                                line = line.rstrip()
                                if line.strip() != "":
                                        a_line = line.split(' ')
                                        name = a_line[0]
                                        registration_no = a_line[1]

                                        filename = self.CERT_DIR + name + "_" + registration_no  + "_public"
                                        if os.path.isfile(filename) :
                                                try:
                                                        f = open(filename,'r')
                                                        public_key = rsa.importKey(f.read())

                                                        # insert in database
                                                        self.put(name,registration_no,public_key)
                                                except Exception as e:
                                                        print e
                                        else:
                                                print "EXCEPTION ! File " + filename + " not found"
                else:
                        print "Voter File not found " + self.voter_file

                # initialize election candidates	
		index = 0
                for candidate in self.election_candidates:
			index = index + 1
                        self.election_database[candidate] = [index, 0]

        def put(self,name,registration_no, public_key):
                voting_status = False
                voted_candidate = None
		voting_time = time.asctime( time.localtime(time.time()) )
                self.voter_database[registration_no] = [name, public_key, voting_status, voted_candidate, voting_time]


        def get(self,registration_no):
                return self.voter_database[registration_no]

        def update(self,registration_no, voting_status):
                self.voter_database[registration_no][2] = voting_status


        def validate_voter(self,name,registration_no):
                if registration_no in self.voter_database:
                        if self.voter_database[registration_no][0] == name:
                                return True
                        else:
                                return False
                else:
                        return False

        def electionVote(self,registration_no,index):

                if self.voter_database[registration_no][2] == True:
                        # has already voted
                        return False
                else:
			elec_candidate = None
			for candidate in self.election_database:
				if self.election_database[candidate][0] == index:
					self.election_database[candidate][1] = self.election_database[candidate][1] + 1
					elec_candidate = candidate
					break
			if elec_candidate == False:
				return False
			
                        self.voter_database[registration_no][2] = True
                        self.voter_database[registration_no][3] = elec_candidate
			voting_time = time.asctime( time.localtime(time.time()) )
			self.voter_database[registration_no][4] = voting_time

			if self.write_to_storage == True:
				if not os.path.isdir(self.persistent_storage):
					# create the directory
		                        os.makedirs(self.persistent_storage)

				filename = self.persistent_storage + self.result_file
				file_handle = open(filename, "w+")
				for candidate in self.election_database:
					buffer = candidate + " \t " + str(self.election_database[candidate][1]) + "\n"
					file_handle.write(buffer)
				file_handle.close()

				filename = self.persistent_storage + self.history_file
				file_handle = open(filename, "w+")
				for reg_no in self.voter_database:
					if self.voter_database[reg_no][2] == True:
						buffer = str(reg_no) + " \t " + str(self.voter_database[reg_no][4]) + "\n"
						file_handle.write(buffer)
				file_handle.close()

			return True

        def get_election_result(self):

                # check if all voters have voted in election
                for reg_no in self.voter_database:
                        if self.voter_database[reg_no][2] == False:
                                return "False"

		# print the result

		max = 0	
		# find candidate with maximum maximum votes	
		for candidate in self.election_database:
			if int(self.election_database[candidate][1]) > max:
				max = int(self.election_database[candidate][1])
		
		winner = ""
		for candidate in self.election_database:
			if int(self.election_database[candidate][1]) == max:	
				winner = winner + candidate + " "
	
		buffer = winner + " Win \n"
		resultant = buffer
		for candidate in self.election_database:
               		buffer = candidate + " \t " + str(self.election_database[candidate][1]) + "\n"
			resultant = resultant + buffer
		
		return resultant

	def get_voter_history(self,reg_no):

		resultant = ""
              	if self.voter_database[reg_no][2] == True:
                    	buffer = str(reg_no) + " \t " + str(self.voter_database[reg_no][4]) + "\n"
			resultant = resultant + buffer
		
		return resultant

	def hasVoted(self,registration_no):
		if self.voter_database[registration_no][2] == True: 
			return True
		else:
			return False



Voting_Crypto.py
#!/usr/bin/python

from base64 import b64encode, b64decode
import rsa

class Voting_Crypto:
	CERT_DIR = "certs/"

	##################################################
	# loads public cert and returns
	def load_public(self, certfile):
		certfile = self.CERT_DIR + "/" + certfile
		f = open(certfile,'r')
       		key = rsa.importKey(f.read())
		return key

	##################################################
	# Loads private cert and returns
	def load_private(self, certfile):
		certfile = self.CERT_DIR + "/" + certfile
		f = open(certfile,'r')
                key = rsa.importKey(f.read())                                               
                return key

	###################################################
	# Encrypt plain text using public key of Voting server
	# Sign the message using private key of voting client
	def encrypt_message(self, public_vf, private_cl, name, reg_no, stage, extension):
		reg_no = str(reg_no)
		stage = str(stage)
		extension = str(extension)

		try:	
			# E(pub(VF), name||vnumber)||DS(name)
		
			# name || vnumber
			plaintex = name + ":" + reg_no + ":" + stage + ":" + extension 
		
			# encrypt, E(pub(VF), name||vnumber)
			ciphertext = b64encode(rsa.encrypt(plaintex, public_vf))
	
			# sign name, DS(name), signed by voter clients private key
			signature = b64encode(rsa.sign(name, private_cl, "SHA-512"))

			# concatinate and return
			return ciphertext + signature

		except Exception as e:
			print "EXCEPTION ! Invalid key, encrypt_message"
                        print e
                        return False

	#####################################################
	# Decrypt message using private key of voting server
	# Verify signature using public key of client
	def decrypt_message(self, private_vf, ciphertext):

		try:
			# break message into parts, 1st part is cipher text next is the signature
			signature = ciphertext[344:344+344]
			ciphertext = ciphertext[0:344]
		
			# decrypt cipher text
			plaintext = rsa.decrypt(b64decode(ciphertext), private_vf)	
	
			plaintext = plaintext.split(":")
			
			name = plaintext[0]
			reg_no = plaintext[1]
			stage = plaintext[2]
			extension = plaintext[3]
			
			public_cl = self.load_public(name + "_" + reg_no + "_" + "public")

			# verify signature first
			verify = rsa.verify(name, b64decode(signature), public_cl)
	
			if verify == False:
				return (False,name,reg_no,stage,extension)

			return (True,name,reg_no,stage,extension)

		except Exception as e:
			print "EXCEPTION ! Invalid key, decrypt_message"
			print e
			return (False,'','','','')

'''
if __name__ == "__main__":
	obj = Voting_Crypto ()
	name = "Alice"
	reg_no = "112550000"

	public_vf = obj.load_public("voter_server_public")
       	private_vf = obj.load_private("voter_server_private")
	
	private_cl = obj.load_private(name + "_" + reg_no + "_" + "private")
	
	cip =  obj.encrypt_message(public_vf, private_cl, name, reg_no,"")

	print cip
	plain = obj.decrypt_message(private_vf,cip)
	print plain
'''
